### PR TITLE
[MOSIP-25631] Added reusable workflows

### PR DESCRIPTION
**Pull request by syedsalman3753**

### CODE REVIEW

**Review by gsasikumar — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-build.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+          mkdir -p "${{ inputs.SERVICE_LOCATION }}"
+          unzip "./${{ inputs.BUILT_ARTIFACT }}.zip" -d "${{ inputs.SERVICE_LOCATION }}"
+
+      - name: Get current date
+        run: echo "BUILD_TIME=$(date +'%Y-%m-%d-%H-%M-%S')"  >> $GITHUB_ENV
+
+      - name: Build image
+        run: |
+          cd "./${{env.SERVICE_LOCATION}}"
+          docker build . --build-arg SOURCE=mosip --build-arg COMMIT_HASH=$(git rev-parse HEAD) --build-arg COMMIT_ID=$(git rev-parse --short HEAD) --build-arg BUILD_TIME=$BUILD_TIME --file Dockerfile --tag ${{ env.SERVICE_NAME }}
```
[the remark this answers is not in the document]
Comment (ckm007) [reply .github/workflows/docker-build.yml:73]: All the actiones performed by the reusable workflow applies on the caller workflow.
So, commit has also will be taken from caller. @syedsalman3753 then also please verify once before resolving the conversation.

File: .github/workflows/helm-lint.yml [hunk 2]
```diff
@@ -0,0 +1,135 @@
+name: Lint and Test Charts
```
[the remark this answers is not in the document]
Comment (ckm007) [reply .github/workflows/helm-lint.yml:1]: channel should be internal devops and devops.

**Comment by ckm007** [comment]
for all actions lets share a message to slack channel upon failures.
Make sure on failure slack is also configurable.
To differentiate in case of below mentioned repos share the failure message in both techleads as well as devops.
1. mosip-infra.
2. mosip-helm.
3. mosip-onboarder
4. postgres-init
5. keycloak-init
6. mosip-data
7. data-srchive

**Comment by ckm007** [comment]
Also lets add one workflow for one type of job. In the same job we can handle multiple callers relates cases in one workflow.
For e.g. Instead of having maven-sonar, gradel-sonar and npm-sonar. Lets have one workflow for sonar and then handle all these based upon condition and environment from there.

File: .github/workflows/docker-build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/docker-build.yml:85]: Commit id are taken from caller workflow

File: .github/workflows/docker-build.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+          fi
+          echo IMAGE_ID=$IMAGE_ID
+          echo VERSION=$VERSION
+          docker tag $SERVICE_NAME $IMAGE_ID:$VERSION
+          docker push $IMAGE_ID:$VERSION
+
+      - uses: 8398a7/action-slack@v3
+        with:
+          status: ${{ job.status }}
+          fields: repo,message,commit,workflow,job # selectable (default: repo,message)
```
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/docker-build.yml]: Added

File: .github/workflows/gradlew-publish-to-npm-registry.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
+name: Publish to NPM
```
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/gradlew-publish-to-npm-registry.yml]: renamed the workflow as requested

File: .github/workflows/npm-sonar-analysis.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+              sonar.projectKey=mosip_admin-ui
+              sonar.organization=${{ secrets.ORG_KEY }}
+              sonar.sourceEncoding=UTF-8
+              sonar.sources=src
+              sonar.exclusions=**/node_modules/**
+              sonar.tests=src
+              sonar.test.inclusions=**/*.spec.ts
+              sonar.typescript.lcov.reportPaths=coverage/lcov.info" >> sonar-project.properties
+        npm install sonar-scanner && npm run sonar
+      env:
```
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/npm-sonar-analysis.yml:53]: Added slack alerts for all workflows

File: .github/workflows/helm-lint-manual.yml [hunk 6]
```diff
@@ -0,0 +1,152 @@
+name: Run Lint and Test Charts manually
```
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/helm-lint-manual.yml:1]: Added slack alerts for all workflows

File: .github/workflows/helm-lint.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/helm-lint.yml:1]: > channel should be internal devops and devops.

@ckm007 I have implemented the alert; kindly ensure that you update the GitHub secret labeled as `SLACK_DEVOPS_WEBHOOK`.

File: .github/workflows/gradlew-publish-to-npm-registry.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        run: cd ${{ inputs.SERVICE_LOCATION }} && npm run verify
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v3.7.0
+        with:
+          package_path: '${{ inputs.SERVICE_LOCATION }}'
+          registry-url: 'https://registry.npmjs.org/'
+
+      - name: Publish package
+        run: cd ${{ inputs.SERVICE_LOCATION }} && npm publish
```
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/gradlew-publish-to-npm-registry.yml]: This is used to publish npm android applications to the npm registry

**Comment by syedsalman3753** [comment]
> for all actions lets share a message to slack channel upon failures. Make sure on failure slack is also configurable. To differentiate in case of below mentioned repos share the failure message in both techleads as well as devops.
> 
> 1. mosip-infra.
> 2. mosip-helm.
> 3. mosip-onboarder
> 4. postgres-init
> 5. keycloak-init
> 6. mosip-data
> 7. data-srchive

Added configuration to slack url

**Comment by syedsalman3753** [comment]
> Also lets add one workflow for one type of job. In the same job we can handle multiple callers relates cases in one workflow. For e.g. Instead of having maven-sonar, gradel-sonar and npm-sonar. Lets have one workflow for sonar and then handle all these based upon condition and environment from there.

@ckm007  As discussed, we will try this in the next round.

File: .github/workflows/gradlew-publish-to-npm-registry.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (syedsalman3753) [reply .github/workflows/gradlew-publish-to-npm-registry.yml]: The developer has confirmed that it is necessary to build the application before its publication on the npm registry.

**Review by ckm007 — APPROVED** [review]

**Review by gsasikumar — APPROVED** [review]

**Pull request merged**